File: langs/ruby/lib/dotlyte/masking.rb
# frozen_string_literal: true

module Dotlyte
  # Sensitive value masking for DOTLYTE v2.
  module Masking
    REDACTED = "[REDACTED]"

    SENSITIVE_PATTERNS = [
      /password/i, /secret/i, /token/i, /api[_-]?key/i,
      /private[_-]?key/i, /access[_-]?key/i, /auth/i,
      /credential/i, /connection[_-]?string/i, /dsn/i,
      /encryption[_-]?key/i, /signing[_-]?key/i, /certificate/i
    ].freeze

    # Build the set of sensitive keys (auto-detected + schema).
    #
    # @param data [Hash] the config data
    # @param schema_keys [Array<String>] keys marked sensitive in schema
    # @return [Set<String>]
    def self.build_sensitive_set(data, schema_keys = [])
      set = Set.new(schema_keys)
      flat_keys = flatten_keys(data)

      flat_keys.each do |key|
        SENSITIVE_PATTERNS.each do |pat|
          if pat.match?(key)
            set.add(key)
            break
          end
        end
      end

      set
    end

    # Redact sensitive values in a deep hash.
    #
    # @param data [Hash]
    # @param sensitive_keys [Set<String>]
    # @return [Hash] copy with redacted values
    def self.redact(data, sensitive_keys, prefix = "")
      data.each_with_object({}) do |(key, value), result|
        full_key = prefix.empty? ? key.to_s : "#{prefix}.#{key}"

        result[key] = if sensitive_keys.include?(full_key)
                        REDACTED
                      elsif value.is_a?(Hash)
                        redact(value, sensitive_keys, full_key)
                      else
                        value
                      end
      end
    end

    # Partially show a value: first 2 chars visible, rest masked.
    def self.format_redacted(value)
      return REDACTED if value.nil?
      return "*" * value.length if value.length <= 4

      value[0..1] + "*" * (value.length - 2)
    end

    # Compile glob patterns into an array of Regexp objects.
    #
    # Supports `*` (any chars except dot) and `**` (any chars including dot).
    #
    # @param patterns [Array<String>] glob-style patterns (e.g., "db.*", "**secret**")
    # @return [Array<Regexp>] compiled regular expressions
    def self.compile_patterns(patterns)
      patterns.map do |pat|
        # Escape everything except our glob wildcards
        re = pat.gsub(".", "\\.") # escape dots first
        re = re.gsub("**", "\x00DOUBLESTAR\x00")
        re = re.gsub("*", "[^.]*")
        re = re.gsub("\x00DOUBLESTAR\x00", ".*")
        Regexp.new("\\A#{re}\\z", Regexp::IGNORECASE)
      end
    end

    # Build a sensitive key set using glob patterns and schema-declared sensitive keys.
    #
    # @param keys [Array<String>] all flat keys to check
    # @param patterns [Array<String>] glob patterns for sensitive key matching
    # @param schema_sensitive [Set<String>] keys explicitly marked sensitive in schema
    # @return [Set<String>] combined set of sensitive keys
    def self.build_sensitive_set_with_patterns(keys, patterns:, schema_sensitive: Set.new)
      compiled = compile_patterns(patterns)
      set = Set.new(schema_sensitive)

      keys.each do |key|
        compiled.each do |re|
          if re.match?(key)
            set.add(key)
            break
          end
        end
      end

      set
    end

    # Create an audit proxy that wraps a Hash and invokes a callback on sensitive key access.
    #
    # @param data [Hash] the configuration data
    # @param sensitive_keys [Set<String>] keys that trigger the callback
    # @param on_access [Proc] callback receiving the accessed key name
    # @return [Object] a frozen proxy object with [] and get methods
    def self.create_audit_proxy(data, sensitive_keys:, on_access:)
      frozen_data = data.freeze
      sens = sensitive_keys.is_a?(Set) ? sensitive_keys : Set.new(sensitive_keys)
      cb = on_access

      proxy = Object.new

      proxy.define_singleton_method(:[]) do |key|
        cb.call(key) if sens.include?(key)
        frozen_data[key]
      end

      proxy.define_singleton_method(:get) do |key, default_val = nil|
        cb.call(key) if sens.include?(key)
        val = frozen_data[key]
        val.nil? ? default_val : val
      end

      proxy.define_singleton_method(:keys) do
        frozen_data.keys
      end

      proxy.define_singleton_method(:to_h) do
        frozen_data.dup
      end

      proxy.freeze
    end

    class << self
      private

      def flatten_keys(data, prefix = "", out = [])
        data.each do |key, value|
          full_key = prefix.empty? ? key.to_s : "#{prefix}.#{key}"
          if value.is_a?(Hash)
            flatten_keys(value, full_key, out)
          else
            out << full_key
          end
        end
        out
      end
    end
  end
end

File: langs/dotnet/src/Dotlyte/Masking.cs
using System.Text.RegularExpressions;

namespace Dotlyte;

/// <summary>
/// Sensitive value masking for DOTLYTE v2.
/// </summary>
public static class Masking
{
    /// <summary>The replacement string for redacted values.</summary>
    public const string Redacted = "[REDACTED]";

    private static readonly Regex[] SensitivePatterns =
    [
        new(@"password", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        new(@"secret", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        new(@"token", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        new(@"api[_\-]?key", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        new(@"private[_\-]?key", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        new(@"access[_\-]?key", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        new(@"auth", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        new(@"credential", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        new(@"connection[_\-]?string", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        new(@"dsn", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        new(@"encryption[_\-]?key", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        new(@"signing[_\-]?key", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        new(@"certificate", RegexOptions.IgnoreCase | RegexOptions.Compiled),
    ];

    /// <summary>
    /// Build the set of sensitive keys (auto-detected + schema).
    /// </summary>
    public static List<string> BuildSensitiveSet(
        Dictionary<string, object?> data,
        IEnumerable<string>? schemaKeys = null)
    {
        var set = new HashSet<string>(schemaKeys ?? Enumerable.Empty<string>());

        foreach (var key in FlattenKeys(data))
        {
            foreach (var pattern in SensitivePatterns)
            {
                if (pattern.IsMatch(key))
                {
                    set.Add(key);
                    break;
                }
            }
        }

        return set.ToList();
    }

    /// <summary>
    /// Redact sensitive values in a deep dictionary.
    /// </summary>
    public static Dictionary<string, object?> RedactDictionary(
        Dictionary<string, object?> data,
        IReadOnlyCollection<string> sensitiveKeys,
        string prefix = "")
    {
        var result = new Dictionary<string, object?>();
        foreach (var (key, value) in data)
        {
            var fullKey = string.IsNullOrEmpty(prefix) ? key : $"{prefix}.{key}";

            if (sensitiveKeys.Contains(fullKey))
                result[key] = Redacted;
            else if (value is Dictionary<string, object?> nested)
                result[key] = RedactDictionary(nested, sensitiveKeys, fullKey);
            else
                result[key] = value;
        }
        return result;
    }

    /// <summary>
    /// Partially show a value: first 2 chars visible, rest masked.
    /// </summary>
    public static string FormatRedacted(string? value)
    {
        if (value is null)
            return Redacted;
        if (value.Length <= 4)
            return new string('*', value.Length);
        return value[..2] + new string('*', value.Length - 2);
    }

    private static List<string> FlattenKeys(
        Dictionary<string, object?> data, string prefix = "")
    {
        var result = new List<string>();
        foreach (var (key, value) in data)
        {
            var fullKey = string.IsNullOrEmpty(prefix) ? key : $"{prefix}.{key}";
            if (value is Dictionary<string, object?> nested)
                result.AddRange(FlattenKeys(nested, fullKey));
            else
                result.Add(fullKey);
        }
        return result;
    }

    /// <summary>
    /// Compile an array of regex pattern strings into <see cref="Regex"/> objects.
    /// </summary>
    /// <param name="patterns">Array of regex pattern strings.</param>
    /// <returns>Array of compiled <see cref="Regex"/> instances.</returns>
    public static Regex[] CompilePatterns(string[] patterns)
    {
        return patterns
            .Select(p => new Regex(p, RegexOptions.IgnoreCase | RegexOptions.Compiled))
            .ToArray();
    }

    /// <summary>
    /// Build a set of sensitive keys by combining explicit keys, regex patterns, and schema-defined sensitive keys.
    /// </summary>
    /// <param name="keys">Explicit key names of configuration entries to check against patterns.</param>
    /// <param name="patterns">Optional regex patterns to match against key names.</param>
    /// <param name="schemaSensitive">Optional set of keys marked sensitive in a schema.</param>
    /// <returns>A <see cref="HashSet{T}"/> of all keys determined to be sensitive.</returns>
    public static HashSet<string> BuildSensitiveSetWithPatterns(
        string[] keys,
        string[]? patterns = null,
        HashSet<string>? schemaSensitive = null)
    {
        var result = new HashSet<string>(schemaSensitive ?? Enumerable.Empty<string>());

        var compiledPatterns = patterns is not null
            ? CompilePatterns(patterns)
            : SensitivePatterns;

        foreach (var key in keys)
        {
            foreach (var pattern in compiledPatterns)
            {
                if (pattern.IsMatch(key))
                {
                    result.Add(key);
                    break;
                }
            }
        }

        return result;
    }
}

/// <summary>
/// An audit proxy that wraps a configuration dictionary and invokes a callback
/// whenever a sensitive key is accessed through the indexer.
/// </summary>
public sealed class ConfigAuditProxy
{
    private readonly Dictionary<string, object?> _data;
    private readonly HashSet<string> _sensitiveKeys;
    private readonly Action<string> _onAccess;

    /// <summary>
    /// Initializes a new <see cref="ConfigAuditProxy"/>.
    /// </summary>
    /// <param name="data">The backing configuration data.</param>
    /// <param name="sensitiveKeys">Set of keys considered sensitive.</param>
    /// <param name="onAccess">Callback invoked with the key name when a sensitive key is accessed.</param>
    public ConfigAuditProxy(
        Dictionary<string, object?> data,
        HashSet<string> sensitiveKeys,
        Action<string> onAccess)
    {
        _data = data ?? throw new ArgumentNullException(nameof(data));
        _sensitiveKeys = sensitiveKeys ?? throw new ArgumentNullException(nameof(sensitiveKeys));
        _onAccess = onAccess ?? throw new ArgumentNullException(nameof(onAccess));
    }

    /// <summary>
    /// Access a configuration value by key. Triggers the audit callback for sensitive keys.
    /// </summary>
    /// <param name="key">The configuration key.</param>
    /// <returns>The configuration value, or <c>null</c> if not found.</returns>
    public object? this[string key]
    {
        get
        {
            if (_sensitiveKeys.Contains(key))
            {
                _onAccess(key);
            }

            _data.TryGetValue(key, out var value);
            return value;
        }
    }

    /// <summary>
    /// Get all keys in the backing dictionary.
    /// </summary>
    public IReadOnlyCollection<string> Keys => _data.Keys;
}
